packages/coil-extension/src/background/services/Stream.ts

```diff
@@ -116,133 +113,158 @@ export class Stream extends EventEmitter {
     }
   ) {
     super()
-
     this._paymentPointer = paymentPointer
     this._requestId = requestId
     this._spspUrl = spspEndpoint
     this._authToken = token
-    this._tiers = container.get(BandwidthTiers)
-    this._coilDomain = container.get(tokens.CoilDomain)
     this._anonTokens = container.get(AnonymousTokens)
-
-    this._assetCode = ''
-    this._assetScale = 0
-    this._exchangeRate = 1
-
-    this._active = false
-    this._looping = false
-    this._attempt = null
-    this._lastDelivered = 0
     this._initiatingUrl = initiatingUrl
+    this.loop = new StreamLoop({
+      tabLogger: this.tabLogger,
+      logger: this._debug,
+      schedule: this._schedule
+    })
+    this.loop.on('run:error', (_err: Error) => (this._paying = false))
 
-    const server = new URL(this._coilDomain)
+    const server = new URL(container.get(tokens.CoilDomain))
     server.pathname = '/btp'
     this._server = server.href.replace(/^http/, 'btp+ws')
-
     if (BTP_ENDPOINT) {
       this._server = BTP_ENDPOINT
     }
   }
 
-  async start() {
-    if (this._active) return
-    this._active = true
-
-    // reset this upon every start *before* early exit while _looping
+  async start(): Promise<void> {
+    // reset this upon every start *before* early exit while looping
     this._packetNumber = 0
+    this.isPaused = false
+    await this.loop.run(async (tokenFraction: number): Promise<Connection> => {
```

Comment:
@sentientwaffle 
It seems for some reason that this attempt callback passed in to run(...) is getting executed twice in quick succession. I don't recall seeing that before your latest changes, but I may have missed it ?

Reply:
Looking closer, I think it's the winding down of the previous stream.
When you refresh the page, it seems like it's trying to wind down 

Question: does it try to post pay in this situation?

Reply:
It seems it does, and successfully (at least the times I've looked)

Reply:
It was somewhat surprising that it seems to be connecting again, but that's working as intended, as the connection is discarded in between batches. 

So when you refresh the tab, it will close the stream, which will in turn put the stream in Ending state, and pay out whatever is necessary. This will (may?) require creating a new connection.